test/gtest/uct/ib/test_cqe_zipping.cc

```diff
@@ -0,0 +1,174 @@
+                zipped_cqes_cnt += UCS_STATS_GET_COUNTER(stats, counter);
+            }
+        }
+        return zipped_cqes_cnt;
+    }
+
+    virtual void init()
+    {
+        stats_activate();
+        modify_config("IB_CQE_ZIPPING_ENABLE", "y");
+
+        test_uct_ib_with_specific_port::init();
+
+        peers.emplace_back(*this);
+        peers.emplace_back(*this);
+
+        peers[0].m_entity->connect(0, *peers[1].m_entity, 0);
+        peers[1].m_entity->connect(0, *peers[0].m_entity, 0);
+
+        flush();
+    }
+
+    void flush() {
+        uct_test::flush();
+        send_cnt = recv_cnt = 0;
+    }
+
+private:
+    struct peer {
+        peer(test_cqe_zipping &test) :
+            m_test(test),
+            m_entity(test.create_entity(0)),
+            m_buf(new mapped_buffer(buf_size, 0, *m_entity))
+        {
+            m_test.m_entities.push_back(m_entity);
+            uct_iface_set_am_handler(m_entity->iface(), 0, am_cb, &m_test, 0);
+        }
+
+        test_cqe_zipping               &m_test;
+        entity                         *m_entity;
+        std::unique_ptr<mapped_buffer> m_buf;
+    };
+
+    virtual void cleanup()
+    {
+        flush();
+        peers.clear();
+
+        uct_test::cleanup();
+        stats_restore();
+    }
+
+    ucs_status_t am_zcopy(peer &input_peer)
+    {
+        auto &entity  = input_peer.m_entity;
+        auto &sendbuf = input_peer.m_buf;
+
+        size_t size = ucs_min(entity->iface_attr().cap.am.max_zcopy, buf_size);
+        uct_iov_t iov{(char*)sendbuf->ptr(), size, sendbuf->memh(), 0, 1};
+
+        return uct_ep_am_zcopy(entity->ep(0), 0, sendbuf->ptr(), 0, &iov, 1, 0,
+                               nullptr);
+    }
+
+    static ucs_status_t
+    am_cb(void *arg, void *data, size_t length, unsigned flags)
+    {
+        test_cqe_zipping *test = static_cast<test_cqe_zipping*>(arg);
+        ++test->recv_cnt;
+        return UCS_OK;
+    }
+
+    ucs_stats_node_t *iface_stats(const peer &p)
+    {
+        return ucs_derived_of(p.m_entity->iface(), uct_ib_iface_t)->stats;
+    }
+
+    void check_port_attr() {
+        if (IBV_PORT_IS_LINK_LAYER_ETHERNET(&m_port_attr)) {
+            /* Hard to get zipped CQEs on Ethernet link */
+            UCS_TEST_SKIP_R("skip CQE zipping test for Ethernet link type");
+        }
+    }
+
+    static constexpr size_t buf_size = 2048;
+    std::vector<peer>       peers = {};
+    size_t                  send_cnt = 0, recv_cnt = 0;
+};
+
+/* We test only ZCOPY with 4K size messages due to good PCI load and stably
+ * CQE zipping reproducing on all platforms.
+ */
+UCS_TEST_SKIP_COND_P(test_cqe_zipping, zcopy, RUNNING_ON_VALGRIND)
+{
+    const ucs_time_t deadline = ucs::get_deadline(60);
+
+    /* Some transports (e.g. UD) dynamically extend the amount of sends,
+     * so we must do several send/recv iterations to make CQE zipping
+     * sensable from the FW side.
+     */
+    while (ucs_get_time() < deadline) {
+        flush();
```

Comment:
why need to flush?

Reply:
Without that flush DC and RC transports successfully execute send command several times only during the first `send_while_possible` call and then always return `UCS_ERR_NO_RESOURCE` at the first iteration of send loop.

Reply:
ok, pls comment it in the code